tests/unstable-fn.egg

```diff
@@ -0,0 +1,65 @@
+(datatype Math
+  (Num i64)
+  (Var String)
+  (Add Math Math)
+  (Mul Math Math))
+
+(rewrite (Mul (Num x) (Num y)) (Num (* x y)))
+
+(datatype MathList
+  (Nil)
+  (Cons Math MathList))
+
+(sort MathFn (UnstableFn (Math) Math))
+
+
+(function square (Math) Math)
+(rewrite (square x) (Mul x x))
+
+(let square-fn (unstable-fn "square" ))
+
+;; test that we can call a function
+(let squared-3 (unstable-app square-fn (Num 3)))
+(check (= squared-3 (square (Num 3))))
+
+;; test that we can apply a function to a list
+
+(function list-map-math (MathList MathFn) MathList)
+(rewrite (list-map-math (Nil) fn) (Nil))
+(rewrite (list-map-math (Cons x xs) fn) (Cons (unstable-app fn x) (list-map-math xs fn)))
+
+(let x (Cons (Num 1) (Cons (Num 2) (Cons (Num 3) (Nil)))))
+(let squared-x (list-map-math x square-fn))
+(run 100)
+(check (= squared-x (Cons (Num 1) (Cons (Num 4) (Cons (Num 9) (Nil))))))
+
+;; Test that we can partially apply a function in a rewrite rule
+
+(function list-multiply-by (MathList Math) MathList)
+(rewrite (list-multiply-by l i) (list-map-math l (unstable-fn "Mul" i)))
+
+(let doubled-x (list-multiply-by x (Num 2)))
+(run 100)
+(check (= doubled-x (Cons (Num 2) (Cons (Num 4) (Cons (Num 6) (Nil))))))
+
+;; Test we can define a higher order compose function
+
+(function composed-math (MathFn MathFn Math) Math)
+(rewrite (composed-math f g v) (unstable-app f (unstable-app g v)))
+
+(let square-of-double (unstable-fn "composed-math" square-fn (unstable-fn "Mul" (Num 2))))
```

Comment:
We shouldn't use strings here for the identifier `composed-math`

Reply:
AFAIK there isn't a way to use real identifiers without a large change to parsing/type checking, which I was trying to avoid for this sort of unstable feature.

Reply:
okay!